db/db_iter.cc

```diff
@@ -267,6 +273,7 @@ class DBIter final: public Iterator {
   }
 
   inline void ResetInternalKeysSkippedCounter() {
+    local_stats_.skip_count_ += num_internal_keys_skipped_;
```

Comment:
I believe `ResetInternalKeysSkippedCounter` is called exactly once per found user key (we call it at the top of all the functions that iterate to somewhere else and also in destructor). Would it be cleaner to decrement by one here? (sorry for suggesting using `next_found_count_` to decrement last time -- I didn't think of this then)

Reply:
to handle the first call to `ResetInternalKeysSkippedCounter ` when you don't want to decrement by one, you could check `!valid_` to detect it. That'd also cause us to not decrement skip count for keys that had errors, which sounds good too.

Reply:
num_internal_keys_skipped_ always has the correct count of skipped internal keys, as far as I understand. I may be missing something as the iterator code is too complicated, but why there is a need to decrement by one for the current case? 

If you look at the definition of `TooManyInternalKeysSkipped()`, and the places it is called from, num_internal_keys_skipped_ seems to incremented correctly ... only when absolutely needed. 
https://github.com/facebook/rocksdb/blob/0996e140bd4da46ce46049c9fd1eae61f3541dee/db/db_iter.cc#L1093-L1103

Reply:
Spoke too soon, without investigating. Man, the iterator code is so tangled!
Since TooManyInternalKeysSkipped is called even in the case where a valid key is found (since it is called at the top of each caller), num_internal_keys_skipped_ should be decremented by 1 for getting the right value. 

I agree with your suggestion to decrement by 1  here, for returning the right skip_count_ stat.

Reply:
My intuition was that if we call it number of keys skipped, we should not count keys that are returned to the user. For example, `FindNextUserEntryInternal` calls `TooManyInternalKeysSkipped` with `increment=true` near the top of its big loop before considering whether to return the key to the user. I think the final case, where it finds a valid key and returns, shouldn't count as a skip.

Reply:
That does sound cleaner. Thanks for the suggestion. We can eliminate the decrements in Seek(), SeekForPrev() and SeekToFirst(). 
The result for the last test case is off by 1 though (expect +7, actual +8) with this approach. What does iter_->SeekForPrev() return - the key with the highest or the lowest sequence number? If its the latter, that would explain it since DBIter would have to skip one to get to the highest sequence number. In my previous implementation, this case was resulting in a double decrement since SeekToLast() can call Prev(), which would increment prev_found_count_.

Reply:
It should return the key with the higher seqnum.

Reply:
The SeekToLast() will see and skip the "b", though, because that's the target of its SeekForPrev on the internal iterator. Maybe that's the cause of one more.

Reply:
I have implemented the decrement in ResetInternalKeysSkipped(). SeekToLast() was calling Prev() directly, which was causing a decrement even though the key is not returned to the user. Fixed that by calling PrevInternal() instead.